pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,308 @@
+        :rtype: tuple(RoutingInfo,
+            dict((ApplicationVertex, str), BaseKeyAndMask))
         """
+        by_app_and_partition_name = \
+            self.__machine_graph.multicast_partitions
         progress = ProgressBar(
-            self.__application_graph.n_vertices, "Allocating routing keys")
+            len(by_app_and_partition_name), "Allocating routing keys")
         routing_infos = RoutingInfo()
-        by_app_vertex = dict()
-        app_mask = self.__mask(max_app_keys_bits)
-
-        for source_index, app_vertex in progress.over(
-                enumerate(self.__application_graph.vertices)):
-            app_key = source_index << max_app_keys_bits
-            if app_vertex in key_bits_map:
-                key_bits = key_bits_map[app_vertex]
-                for machine_index, vertex in enumerate(
-                        app_vertex.machine_vertices):
-                    partitions = self.__machine_graph.\
-                        get_outgoing_edge_partitions_starting_at_vertex(vertex)
-                    partition = partitions.peek()
-                    if partition.traffic_type == EdgeTrafficType.MULTICAST:
-                        mask = self.__mask(key_bits)
-                        key = app_key | machine_index << key_bits
-                        key_and_mask = BaseKeyAndMask(base_key=key, mask=mask)
-                        routing_infos.add_partition_info(
-                            PartitionRoutingInfo([key_and_mask], partition))
-                by_app_vertex[app_vertex] = BaseKeyAndMask(
-                    base_key=app_key, mask=app_mask)
-
-        return routing_infos, by_app_vertex
+        app_part_index = 0
+        for app_id in progress.over(by_app_and_partition_name):
+            while app_id in self.__fixed_used:
+                app_part_index += 1
+            for partition_name, by_partition_name in \
+                    by_app_and_partition_name[app_id].items():
+                machine_vertices = list(by_partition_name)
+                machine_vertices.sort(key=lambda x: x.vertex_slice.lo_atom)
+                if self.__flexible:
+                    n_bits_atoms = self.__atom_bits_per_app_part[
+                        (app_id, partition_name)]
+                    n_bits_machine = self.__n_bits_atoms_and_mac - n_bits_atoms
+                else:
+                    n_bits_atoms = self.__atom_bits_per_app_part[
+                        (app_id, partition_name)]
+                    if n_bits_atoms < self.__n_bits_atoms:
+                        # Ok it fits use global sizes
+                        n_bits_atoms = self.__n_bits_atoms
+                        n_bits_machine = self.__n_bits_machine
+                    else:
+                        # Nope need more use that so adjust n_bits_machine down
+                        n_bits_machine = \
+                            self.__n_bits_atoms_and_mac - n_bits_atoms
```

Comment:
As above, I'm not sure how this makes it global

Reply:
commented added to say in this case it is not global!